### ISSUE TITLE
Barba.js and ScrollMagic conflicts

### ISSUE THREAD

**Issue by nexnetpl** [janpaepke/ScrollMagic#853]
Hi, it is a simple page using both plugins, somehow ScrollMagic stop working during the navigation process
Is it a known issue?

Example http://serwer1853007.home.pl/www/index.html

Thank you

**Comment by dngraphisme** [comment]
Hello, i have a similar problem. Have you solved your problem?

Thanks

**Comment by geanfrancovolpe** [comment]
Have you solved the problem?

**Comment by dngraphisme** [comment]
@janpaepke Please help us for this...

With ajax scrollmagic dont work :(

Thanks for your help !

**Comment by geanfrancovolpe** [comment]
> @janpaepke Please help us for this...
> 
> With ajax scrollmagic dont work :(
> 
> Thanks for your help !

Bro it does work, make a fresh npm install of scroll magic, and import this:

**Comment by geanfrancovolpe** [comment]
> Hi, it is a simple page using both plugins, somehow ScrollMagic stop working during the navigation process
> Is it a known issue?
> 
> Example http://serwer1853007.home.pl/www/index.html
> 
> Thank you

You need to create a `ScrollMagic.scene` to every module, one for `home` and one for `about` because, it doesn't assign the scene to the entire namespaces `divs`. And there's where you need to use `Barba Views`, to init every module `beforeAppear`, or `beforeEnter`. Checkout [Barba V2 Views](https://barba.js.org/docs/v2/user/core.html#view-object) for more information.

**Comment by dngraphisme** [comment]
> > Hi, it is a simple page using both plugins, somehow ScrollMagic stop working during the navigation process
> > Is it a known issue?
> > Example http://serwer1853007.home.pl/www/index.html
> > Thank you
> 
> @nexnetpl You need to init`ScrollMagic.scene` in every module, one for `home` and one for `about` because, it doesn't assign the scene to the entire namespaces `divs`. In the `barba.init()` function declare views to init every module `beforeAppear`, or `beforeEnter` the namespace. Checkout [Barba V2 Views](https://barba.js.org/docs/v2/user/core.html#view-object) for more information.

Hello gean, thanks for your reply.

In fact you put your init in each pages (seen in barbajs) but you do not destroy anything when you leave the current page or when you come back?

You could link me an example that I'm trying to see what you did. I'm also on the V2, the V1 is a little outdated.

I'm back with SM and I'll let you know.

Thank you very much for your return, it helps a lot.

**Comment by geanfrancovolpe** [comment]
@dngraphisme No, i'm declaring `barba.init()` once in a `main.ts`. And creating `homepage` init module apart and importing to `main module`.  See this simple example (maybe not a best practice but it works): 

In main.ts i've declare: 

```

import barba from '@barba/core';
import { initAboutPage , destroyAboutPage } from './about'; //call the about module functions or create an export class, whatever you want.

let transitions = [ //Whatever transitions you'll want to make. e.g. a transition from `home` to `about`.  
    { 
      name: 'loaderTransition',
      sync: false,
      from: { namespace: ['home'] },
      to: { namespace: ['about'] },
      beforeLeave() { //here return a promise before leave home. }, 
      afterEnter() { //here declare functons after enter transition on about namespace. }
     }, 
     {  } //You can declare more than one transition the same way we declare it before, e.g. this time from `about` to `contact`, etc. 
] 

let views = [ 
    { 
       namespace: 'home'
       afterEnter( data ) { 
            initAboutPage(); //for example.
       }, 
       beforeLeave( data ) {
          destroyAboutPage(); //for example
      },
    } //you can create more than one view too.
]

barba.init({ //barba init.
    transitions : transitions,
    views : views
});
```

Then in the about.ts page module you should export these two functions `initAboutPage` and `destroyAboutPage`, it works to me.

**Comment by dngraphisme** [comment]
> afterEnter( data ) { 
>             initAboutPage(); //for example.
>        }, 
>        beforeLeave( data ) {
>           destroyAboutPage(); //for example
>       },

Hello dude, thanks a lot for your help.

I explain to you what I did:

In my components.js (that's where I put all my functions) I created 2 functions:

`function animateADD() {

    var ctrl = new ScrollMagic.Controller();

    $('.animate').each(function () {

        var tl = new TimelineMax();

        TweenMax.set('.pr-bg', {
            left: "0"
        });

        tl.to('.pr-bg', 0.8, {
            force3D: true,
            left: "100%",
            ease: Power2.easeOut
        });

        var scene = new ScrollMagic.Scene({
            triggerElement: this,
            reverse: true,
        })
            .setTween(tl)
            .addIndicators()
            .addTo(ctrl);

    });
};`

**Comment by geanfrancovolpe** [comment]
@dngraphisme a kind of, it does work to you? you can init the controller outside the functions, make it global in your module `components.js`. In that way when you call `destroyAnimateADD()`, it should destroy that global `component.js` controller. The `beforeEnter()` barba views hook, in my app isn't triggering when a page transition (barba transition) starts and end in that namespace, but when the page is loaded directly from the browser, it does work. The others view hooks works well ( `afterEnter()` and `beforeLeave()` ) when a page transition ends in that namespace. Let me know if your code works.

**Comment by dngraphisme** [comment]
> @dngraphisme a kind of, it does work to you? you can init the controller outside the functions, make it global in your module `components.js`. In that way when you call `destroyAnimateADD()`, it should destroy that global `component.js` controller. The `beforeEnter()` barba views hook, in my app isn't triggering when a page transition (barba transition) starts and end in that namespace, but when the page is loaded directly from the browser, it does work. The others view hooks works well ( `afterEnter()` and `beforeLeave()` ) in every case (page load, or page transition). Let me know if your code works.

Hello @geanfrancovolpe ,

Thanks again.

Here is what I did>

1. Components.js

view-source: https: //dev.amourdedieu.fr/static/js/components.js

I created a global function putting scrollmagic in

`function createOSScene ($ el, tl) {};`

In my function : `function animateADD () {}`

I included this>

`createOSScene($ el, tl)`

It starts well in front but when I change page it does not reinvent. (the triggers go away)

-----
In transition I initiated scrollmagic in >

`barba.hooks.beforeEnter (() => {
        PageLoad ();
        initPage ();
        createOSScene (); // here
    });`

then in my sight >

`namespace: 'homepage',
                beforeEnter (data) {
                    home page();
                    parallaxADD ();
                    //console.log (HP 'charging);
                }
                afterEnter (data) {
                    animateADD ();
                }
                beforeLeave (data) {
                    destroyAnimateADD ();
                }`
view-source: https: //dev.amourdedieu.fr/static/js/transition.js

But if I change page as stated above, scrollamgic does not reinvent itself.

**Comment by geanfrancovolpe** [comment]
@dngraphisme I meaned `view methods` instead of `view hooks`. You should use hooks if you want to init some transition in every namespace. But in your case, only by calling a function in `barba.views`, it should work. e.g:

```
namespace: 'homepage', 
beforeEnter (data) { //only works when loading the page directly from the browser but not when transition starts to homepage namespace. I don't recommed it to use. Maybe is a barba bug. 
     homepage(); 
     parallaxADD (); 
}, 
afterEnter (data) { //this callback will wait for a transition end ( if you have a transition declared )
     //homepage(); //if the scene is not too big, it should init faster here.
     //parallaxADD (); //if the scene is not too big, it should init faster here too.
     animateADD (); 
},  
beforeLeave (data) { 
     destroyHomepageFunctions(); // Homepage init functions you want to detroy.
     destroyParallaxADD(); //if you want to.
     destroyAnimateADD (); 
}
```
If you're creating the controller globally, try [destroying the scene](https://scrollmagic.io/docs/ScrollMagic.Scene.html#destroy) instead of the controller. Leave that `controller` variable in memory and only destroy and init the scenes.

**Comment by clmsftsw** [comment]
Hello,

It does not work for me. It seems like I am not able to reinit ScrollMagic after I destroyed it. Can somebody help?

**Comment by janpaepke** [comment]
Long overdue reply — thanks to @geanfrancovolpe and @dngraphisme for working through the solution in the comments above.

The core issue was that Barba.js replaces DOM content via AJAX, so ScrollMagic scenes created for the old DOM become stale after a page transition. The solution (as discussed above) is to destroy scenes/controller in `beforeLeave` and reinitialize in `afterEnter`. In v3 this pattern is even cleaner: each `new ScrollMagic({ element })` instance is self-contained (no shared Controller), so you just call `.destroy()` on each instance when navigating away and create new ones for the new page content.

---
**ScrollMagic v3 is now available** — a complete rewrite with an entirely new architecture and API. v2 will not receive further updates (2.0.9 is the final release).

👉 `npm install scrollmagic@next` — [Docs](https://github.com/janpaepke/ScrollMagic/blob/main/README.md) · [Release notes](https://github.com/janpaepke/ScrollMagic/releases/tag/v3.0.0-beta.0) · [npm](https://www.npmjs.com/package/scrollmagic/v/3.0.0-beta.0)
